pkg/fuzzer/job.go

```diff
@@ -73,113 +92,146 @@ func (job *triageJob) run(fuzzer *Fuzzer) {
+	if !job.fuzzer.Config.NewInputFilter(callName) {
 		return
 	}
-	fuzzer.Logf(2, "added new input for %v to the corpus: %s", callName, job.p)
 	if job.flags&ProgSmashed == 0 {
-		fuzzer.startJob(fuzzer.statJobsSmash, &smashJob{
-			p:    job.p.Clone(),
-			call: job.call,
+		job.fuzzer.startJob(job.fuzzer.statJobsSmash, &smashJob{
+			p:    p.Clone(),
+			call: call,
 		})
 	}
+	job.fuzzer.Logf(2, "added new input for %v to the corpus: %s", callName, p)
 	input := corpus.NewInput{
-		Prog:     job.p,
-		Call:     job.call,
+		Prog:     p,
+		Call:     call,
 		Signal:   info.stableSignal,
 		Cover:    info.cover.Serialize(),
 		RawCover: info.rawCover,
 	}
-	fuzzer.Config.Corpus.Save(input)
-}
-
-type deflakedCover struct {
-	stableSignal    signal.Signal
-	newStableSignal signal.Signal
-	cover           cover.Cover
-	rawCover        []uint64
+	job.fuzzer.Config.Corpus.Save(input)
 }
 
-func (job *triageJob) deflake(exec func(*queue.Request, ProgFlags) *queue.Result, cover *Cover,
-	stat *stats.Val, rawCover bool) (info deflakedCover, stop bool) {
-	// As demonstrated in #4639, programs reproduce with a very high, but not 100% probability.
-	// The triage algorithm must tolerate this, so let's pick the signal that is common
-	// to 3 out of 5 runs.
-	// By binomial distribution, a program that reproduces 80% of time will pass deflake()
-	// with a 94% probability. If it reproduces 90% of time, it passes in 99% of cases.
-	const (
-		needRuns = 3
-		maxRuns  = 5
-	)
-	signals := make([]signal.Signal, needRuns)
-	signals[0] = job.newSignal.Copy()
-	for i := 1; i < maxRuns; i++ {
-		if job.newSignal.IntersectsWith(signals[needRuns-1]) {
-			// We already have the right deflaked signal.
-			break
+func (job *triageJob) deflake(exec func(*queue.Request, ProgFlags) *queue.Result) (stop bool) {
+	prevTotalNewSignal := 0
+	for run := 1; ; run++ {
+		totalNewSignal := 0
+		indices := make([]int, 0, len(job.calls))
+		for call, info := range job.calls {
+			indices = append(indices, call)
+			totalNewSignal += len(info.newSignal)
 		}
-		if left := maxRuns - i; left < needRuns &&
-			!job.newSignal.IntersectsWith(signals[needRuns-left-1]) {
-			// There's no chance to get coverage common to needRuns.
+		// For fuzzing programs we stop if we already have the right deflaked signal for all calls,
+		// or there's no chance to get coverage common to needRuns for all calls.
+		if job.flags&ProgFromCorpus == 0 {
+			if run >= deflakeMaxRuns {
+				break
+			}
+			haveSignal, noChance := true, true
+			for _, call := range job.calls {
+				if !call.newSignal.IntersectsWith(call.signals[deflakeNeedRuns-1]) {
+					haveSignal = false
+				}
+				if left := deflakeMaxRuns - run; left >= deflakeNeedRuns ||
+					call.newSignal.IntersectsWith(call.signals[deflakeNeedRuns-left-1]) {
+					noChance = false
+				}
+			}
+			if haveSignal || noChance {
+				break
+			}
+		} else if run >= deflakeMaxRuns && prevTotalNewSignal == totalNewSignal {
```

Comment:
Nit: it's de facto `deflakeMinRuns`.

Do we want to have some big `deflakeMaxRuns`, after which we'd abort the process and possibly print `log.Errorf()` to let us know it's happening on syzbot?

Reply:
Done. PTAL.